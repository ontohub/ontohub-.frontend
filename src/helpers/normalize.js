import _ from 'lodash'

/**
 * @param {object} attributes - a valid JSON API attributes object
 * @return {object}
 */
const parseAttributes = (attributes) => {
  return _.mapKeys(attributes, _.rearg(_.camelCase, [1, 0]))
}

/**
 * @param {object} relationships - a valid JSON API relationships object
 * @return {object}
 */
const parseRelationships = (relationships) => {
  let mappedKeys = _.mapKeys(
    relationships,
    (val, key) => `_${_.camelCase(key)}`
  )
  let mappedValues = _.mapValues(mappedKeys, (val) => {
    let data = _.get(val, 'data')
    if (Array.isArray(data)) {
      return _.map(data, (d) => d.id)
    } else {
      return _.get(data, 'id')
    }
  })
  return mappedValues
}

/**
 * @param {object} json - A valid toplevel JSON API response
 * @return {object}
 */
export const normalize = (json) => {
  let attributes = {},
    relationships = {},
    keys = _.chain(json).keys().intersection(['errors', 'data']).value()

  if(_.size(keys) !== 1) {
    let error = {
      status: "500",
      title: "Invalid JSON API document"
    }
    throw JSON.stringify(error);
  }

  if (json.errors) {
    throw JSON.stringify(_.get(json, 'errors'))
  }

  if (_.has(json, 'data')) {
    let data = _.get(json, 'data'), id = _.get(data, 'id')

    if (_.has(data, 'attributes')) {
      attributes = parseAttributes(_.get(data, 'attributes'))
    }
    if (_.has(data, 'relationships')) {
      relationships = parseRelationships(_.get(data, 'relationships'))
    }

    return { id, ...attributes, ...relationships }
  }
}
