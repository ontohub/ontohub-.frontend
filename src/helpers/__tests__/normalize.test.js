import { normalize } from '../normalize'

describe('normalize', () => {
  let mockResponse

  describe('normal response', () => {
    beforeEach(() => {
      mockResponse = {
        data: {
          type: 'mock_responses',
          id: 'mock-response-1',
          attributes: {
            mock_field: 'mock value',
            anotherField: 1234
          },
          relationships: {
            parentResponse: {
              data: {
                id: 'mock-response-0',
                type: 'mock_responses'
              }
            },
            childResponses: {
              data: [
                {
                  id: 'mock-response-2',
                  type: 'mock_responses'
                }
              ]
            }
          }
        }
      }
    })

    it('matches snapshot', () => {
      expect(normalize(mockResponse)).toMatchSnapshot()
    })

    it('camelCases attribute names', () => {
      expect(normalize(mockResponse).mockField).toEqual('mock value')
    })

    it('extracts attributes', () => {
      expect(normalize(mockResponse).id).toEqual('mock-response-1')
      expect(normalize(mockResponse).anotherField).toEqual(1234)
    })

    it('extracts relationships', () => {
      expect(normalize(mockResponse)._parentResponse).toEqual(
        'mock-response-0'
      )
      expect(normalize(mockResponse)._childResponses).toContain(
        'mock-response-2'
      )
    })

    describe('empty model', () => {
      beforeEach(() => {
        mockResponse = {
          data: {
            id: 'mock-response-1',
            type: 'mock_responses'
          }
        }
      })

      it('matches snapshot', () => {
        expect(normalize(mockResponse)).toMatchSnapshot()
      })

      it('returns only the id', () => {
        expect(Object.keys(normalize(mockResponse))).toEqual(['id'])
        expect(normalize(mockResponse).id).toEqual('mock-response-1')
      })
    })
  })

  describe('error response', () => {
    let errors = [
      {
        status: '404',
        title: 'Resource not found',
        detail: 'The requested resource was not found'
      }
    ]

    beforeEach(() => {
      mockResponse = {
        errors
      }
    })

    it('throws the returned errors', () => {
      expect(() => normalize(mockResponse)).toThrow('Resource not found')
    })
  })

  describe('invalid document', () => {
    it('rejects an empty document', () => {
      mockResponse = {}
      expect(() => normalize(mockResponse)).toThrow(
        'Invalid JSON API document'
      )
    })

    it('rejects data and errors', () => {
      mockResponse = {
        data: {},
        errors: {}
      }
      expect(() => normalize(mockResponse)).toThrow(
        'Invalid JSON API document'
      )
    })

    it('rejects only meta', () => {
      mockResponse = {
        meta: {}
      }
      expect(() => normalize(mockResponse)).toThrow(
        'Invalid JSON API document'
      )
    })
  })
})
