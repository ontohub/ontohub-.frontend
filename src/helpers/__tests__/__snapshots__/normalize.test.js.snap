// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalize normal response empty model matches snapshot 1`] = `
Object {
  "id": "mock-response-1",
}
`;

exports[`normalize normal response matches snapshot 1`] = `
Object {
  "_childResponses": Array [
    "mock-response-2",
  ],
  "_parentResponse": "mock-response-0",
  "anotherField": 1234,
  "id": "mock-response-1",
  "mockField": "mock value",
}
`;
